### PR TITLE
Scattergl selection / clustering fix

### LINKED ISSUE #2334: Missing data in scattergl plot when I go juste above 10000 points (regression)
Hi,

I just updated plotly.js from 1.30 to the latest 1.33, and my plot is broken.

I’m rendering about 75000 points with a scattergl plot, and after the update, it looks like most of the points are not rendered at all. What is odd though, is that when I hover over the plot, I can still see popups showing when my mouse is over places where points are supposed to be.

After trying different things, I noticed everything is rendering fine up to 10000 data points. If I go just above, say 10001, my plot is broken and most of the points are not rendered!

Everything worked fine with plotly.js 1.30.

I made two versions of my code, one with 10000 points, and the other with 10001. See for yourself: https://0.x2a.yt/other/private/plotly-test/10000.html
https://0.x2a.yt/other/private/plotly-test/10001.html

(This bug report is a copy/paste from the following forum topic: https://community.plot.ly/t/regression-missing-data-in-scattergl-plot-when-i-go-juste-above-10000-points/8152)

### LINKED ISSUE #2457: scattergl throws console errors on safari
Looks like this issue was introduced in the `scattergl` rewrite (i.e. v1.32.0 doesn't throw any errors) 
https://codepen.io/cpsievert/pen/WzNvVN

screen shot 2018-03-08 at 9 31 52 am

Possibly related to https://github.com/ropensci/plotly/issues/1214

### LINKED ISSUE #2543: Safari not rendering scattergl traces
When I look at [this](https://plot.ly/python/webgl-vs-svg/) example, the first 2 scatterplots are empty.

this is the log :
```
[Error] WebGL: INVALID_VALUE: bufferData: size == 0
	bufferData
	u (plotlyjs-bundle.c9ee06fe8613.js:1:2698524)
	create (plotlyjs-bundle.c9ee06fe8613.js:1:2698698)
	buffer (plotlyjs-bundle.c9ee06fe8613.js:1:2710935)
	exports (plotlyjs-bundle.c9ee06fe8613.js:1:2601002)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:2490441)
	v (plotlyjs-bundle.c9ee06fe8613.js:1:1948244)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:1952031)
	(anonymous function) (plotlyjs-bundle.c9ee06fe8613.js:1:1858047)
	syncOrAsync (plotlyjs-bundle.c9ee06fe8613.js:1:1795598)
	promiseReactionJob
[Error] WebGL: INVALID_VALUE: bufferData: size == 0
	bufferData
	u (plotlyjs-bundle.c9ee06fe8613.js:1:2698524)
	create (plotlyjs-bundle.c9ee06fe8613.js:1:2698698)
	buffer (plotlyjs-bundle.c9ee06fe8613.js:1:2710935)
	exports (plotlyjs-bundle.c9ee06fe8613.js:1:2601120)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:2490441)
	v (plotlyjs-bundle.c9ee06fe8613.js:1:1948244)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:1952031)
	(anonymous function) (plotlyjs-bundle.c9ee06fe8613.js:1:1858047)
	syncOrAsync (plotlyjs-bundle.c9ee06fe8613.js:1:1795598)
	promiseReactionJob
[Error] WebGL: INVALID_VALUE: bufferData: size == 0
	bufferData
	u (plotlyjs-bundle.c9ee06fe8613.js:1:2698524)
	create (plotlyjs-bundle.c9ee06fe8613.js:1:2698698)
	buffer (plotlyjs-bundle.c9ee06fe8613.js:1:2710935)
	exports (plotlyjs-bundle.c9ee06fe8613.js:1:2601173)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:2490441)
	v (plotlyjs-bundle.c9ee06fe8613.js:1:1948244)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:1952031)
	(anonymous function) (plotlyjs-bundle.c9ee06fe8613.js:1:1858047)
	syncOrAsync (plotlyjs-bundle.c9ee06fe8613.js:1:1795598)
	promiseReactionJob
[Error] WebGL: INVALID_VALUE: bufferData: size == 0
	bufferData
	u (plotlyjs-bundle.c9ee06fe8613.js:1:2698524)
	create (plotlyjs-bundle.c9ee06fe8613.js:1:2698698)
	buffer (plotlyjs-bundle.c9ee06fe8613.js:1:2710935)
	exports (plotlyjs-bundle.c9ee06fe8613.js:1:2623124)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:2490478)
	v (plotlyjs-bundle.c9ee06fe8613.js:1:1948244)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:1952031)
	(anonymous function) (plotlyjs-bundle.c9ee06fe8613.js:1:1858047)
	syncOrAsync (plotlyjs-bundle.c9ee06fe8613.js:1:1795598)
	promiseReactionJob
[Error] WebGL: INVALID_VALUE: bufferData: size == 0
	bufferData
	u (plotlyjs-bundle.c9ee06fe8613.js:1:2698524)
	create (plotlyjs-bundle.c9ee06fe8613.js:1:2698698)
	buffer (plotlyjs-bundle.c9ee06fe8613.js:1:2710935)
	exports (plotlyjs-bundle.c9ee06fe8613.js:1:2623177)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:2490478)
	v (plotlyjs-bundle.c9ee06fe8613.js:1:1948244)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:1952031)
	(anonymous function) (plotlyjs-bundle.c9ee06fe8613.js:1:1858047)
	syncOrAsync (plotlyjs-bundle.c9ee06fe8613.js:1:1795598)
	promiseReactionJob
[Error] WebGL: INVALID_VALUE: bufferData: size == 0
	bufferData
	u (plotlyjs-bundle.c9ee06fe8613.js:1:2698524)
	create (plotlyjs-bundle.c9ee06fe8613.js:1:2698698)
	buffer (plotlyjs-bundle.c9ee06fe8613.js:1:2710935)
	exports (plotlyjs-bundle.c9ee06fe8613.js:1:2623231)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:2490478)
	v (plotlyjs-bundle.c9ee06fe8613.js:1:1948244)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:1952031)
	(anonymous function) (plotlyjs-bundle.c9ee06fe8613.js:1:1858047)
	syncOrAsync (plotlyjs-bundle.c9ee06fe8613.js:1:1795598)
	promiseReactionJob
[Error] WebGL: INVALID_VALUE: bufferData: size == 0
	bufferData
	u (plotlyjs-bundle.c9ee06fe8613.js:1:2698524)
	create (plotlyjs-bundle.c9ee06fe8613.js:1:2698698)
	buffer (plotlyjs-bundle.c9ee06fe8613.js:1:2710935)
	exports (plotlyjs-bundle.c9ee06fe8613.js:1:2623284)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:2490478)
	v (plotlyjs-bundle.c9ee06fe8613.js:1:1948244)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:1952031)
	(anonymous function) (plotlyjs-bundle.c9ee06fe8613.js:1:1858047)
	syncOrAsync (plotlyjs-bundle.c9ee06fe8613.js:1:1795598)
	promiseReactionJob
[Error] WebGL: INVALID_OPERATION: drawElements: unable to simulate vertexAttrib0 array
	drawElements
	batch (Anonymous Script 1 (line 1966))
	(anonymous function) (plotlyjs-bundle.c9ee06fe8613.js:1:2692519)
	M (plotlyjs-bundle.c9ee06fe8613.js:1:2632387)
	draw (plotlyjs-bundle.c9ee06fe8613.js:1:2487054)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:2493114)
	v (plotlyjs-bundle.c9ee06fe8613.js:1:1948244)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:1952031)
	(anonymous function) (plotlyjs-bundle.c9ee06fe8613.js:1:1858047)
	syncOrAsync (plotlyjs-bundle.c9ee06fe8613.js:1:1795598)
	promiseReactionJob
[Error] WebGL: INVALID_VALUE: bufferData: size == 0
	bufferData
	u (plotlyjs-bundle.c9ee06fe8613.js:1:2698524)
	create (plotlyjs-bundle.c9ee06fe8613.js:1:2698698)
	buffer (plotlyjs-bundle.c9ee06fe8613.js:1:2710935)
	exports (plotlyjs-bundle.c9ee06fe8613.js:1:2623124)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:2490478)
	v (plotlyjs-bundle.c9ee06fe8613.js:1:1948244)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:1952031)
	(anonymous function) (plotlyjs-bundle.c9ee06fe8613.js:1:1858047)
	syncOrAsync (plotlyjs-bundle.c9ee06fe8613.js:1:1795598)
	promiseReactionJob
[Error] WebGL: INVALID_VALUE: bufferData: size == 0
	bufferData
	u (plotlyjs-bundle.c9ee06fe8613.js:1:2698524)
	create (plotlyjs-bundle.c9ee06fe8613.js:1:2698698)
	buffer (plotlyjs-bundle.c9ee06fe8613.js:1:2710935)
	exports (plotlyjs-bundle.c9ee06fe8613.js:1:2623177)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:2490478)
	v (plotlyjs-bundle.c9ee06fe8613.js:1:1948244)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:1952031)
	(anonymous function) (plotlyjs-bundle.c9ee06fe8613.js:1:1858047)
	syncOrAsync (plotlyjs-bundle.c9ee06fe8613.js:1:1795598)
	promiseReactionJob
[Error] WebGL: INVALID_VALUE: bufferData: size == 0
	bufferData
	u (plotlyjs-bundle.c9ee06fe8613.js:1:2698524)
	create (plotlyjs-bundle.c9ee06fe8613.js:1:2698698)
	buffer (plotlyjs-bundle.c9ee06fe8613.js:1:2710935)
	exports (plotlyjs-bundle.c9ee06fe8613.js:1:2623231)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:2490478)
	v (plotlyjs-bundle.c9ee06fe8613.js:1:1948244)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:1952031)
	(anonymous function) (plotlyjs-bundle.c9ee06fe8613.js:1:1858047)
	syncOrAsync (plotlyjs-bundle.c9ee06fe8613.js:1:1795598)
	promiseReactionJob
[Error] WebGL: INVALID_VALUE: bufferData: size == 0
	bufferData
	u (plotlyjs-bundle.c9ee06fe8613.js:1:2698524)
	create (plotlyjs-bundle.c9ee06fe8613.js:1:2698698)
	buffer (plotlyjs-bundle.c9ee06fe8613.js:1:2710935)
	exports (plotlyjs-bundle.c9ee06fe8613.js:1:2623284)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:2490478)
	v (plotlyjs-bundle.c9ee06fe8613.js:1:1948244)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:1952031)
	(anonymous function) (plotlyjs-bundle.c9ee06fe8613.js:1:1858047)
	syncOrAsync (plotlyjs-bundle.c9ee06fe8613.js:1:1795598)
	promiseReactionJob
[Error] WebGL: INVALID_OPERATION: drawElements: unable to simulate vertexAttrib0 array
	drawElements
	batch (Anonymous Script 2 (line 1966))
	(anonymous function) (plotlyjs-bundle.c9ee06fe8613.js:1:2692519)
	M (plotlyjs-bundle.c9ee06fe8613.js:1:2632387)
	draw (plotlyjs-bundle.c9ee06fe8613.js:1:2487054)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:2493114)
	v (plotlyjs-bundle.c9ee06fe8613.js:1:1948244)
	plot (plotlyjs-bundle.c9ee06fe8613.js:1:1952031)
	(anonymous function) (plotlyjs-bundle.c9ee06fe8613.js:1:1858047)
	syncOrAsync (plotlyjs-bundle.c9ee06fe8613.js:1:1795598)
	promiseReactionJob
[Error] Failed to load resource: the server responded with a status of 404 () (plotly-ico.png, line 0)
```

### LINKED ISSUE #2586: 500k point ScatterGL: unselected points don't fade on second selection
### Overview
When displaing a scattergl trace with a large number of points (> 500k), the second time the selection tool is used the unselected points don't fade into the background.

### To reproduce:
1) Run the following codepen (scattergl with 500k points): https://codepen.io/anon/pen/zjoWpO

2) Perform lasso selection
screen shot 2018-04-27 at 12 46 05 pm

Note that unselected points fade into the background as expeccted

3) Double click to clear selection

4) Perform lasso selection again:
screen shot 2018-04-27 at 12 46 19 pm

Note that now there is no contrast between selected and unselected points

### Comments
 - This doesn't seem to happen with 10k points
 - plotly.js version: 1.36.1
 - Chrome version: Version 65.0.3325.181 (Official Build) (64-bit)

### PR DESCRIPTION
**Pull request by dy**
Fixes #2586 (both issues), fixes #2334, fixes #2543, fixes #2457

Changes:

- added `regl._refresh()` stub to workaround https://github.com/regl-project/regl/issues/474
- fixed point-cluster max depth detection from minimal within dimensions to maximal
- fixed snapping detection
- predefine `regl.buffer` data

image

### CODE REVIEW

**Comment by etpinard** [comment]
Sweet!

@dy can you explain what changes you needed to make to fix the problem?

**Comment by etpinard** [comment]
Amazing. I'll run a few performance tests tomorrow morning to see these changes have a side-effect.

**Comment by etpinard** [comment]
In the meantime @dy, can you confirm that this branch works in IE11?

**Comment by dy** [comment]
Sure
image

**Comment by etpinard** [comment]
Moreover @dy, can you test out the other regl components (regl-line2d, regl-error2d, and sploms)  in Safari?

**Comment by etpinard** [comment]
@dy I'm noticing differences in behavior using 1.36.1 and this branch off https://codepen.io/anon/pen/zjoWpO?editors=0010 

in 1.36.1:

peek 2018-05-01 09-46

this branch:

peek 2018-05-01 09-47

where it looks like 1.36.1 clustered things a little more aggressively making dimmed points appears when selection on a 1e5 pt graph. Is this expected?

**Comment by dy** [comment]
Yes, that is the result of additional levels of clustering rendered due to #2586 mentioned by Alex

**Comment by etpinard** [comment]
In terms of performance, comparing 1.36.1 and this branch:

- pan is slower at 1e5 (nothing too noticeable though),

- but pan is significantly slower at 1e6 points:

1.36.1:

peek 2018-05-01 09-55


this branch:

peek 2018-05-01 09-56

I'd even say panning off this branch perform a little worse than in 1.35.2: https://codepen.io/etpinard/pen/MGpRdW?editors=0010

- no significant slow down during selection comparing 1.36.1 and this branch.

----

So, @dy can you explain why panning is affected by your patches here? Does it have to?

**Comment by etpinard** [comment]
Oh and can you add a mock that lock down the fix for https://github.com/plotly/plotly.js/issues/2334? Thanks!

**Comment by dy** [comment]
I imagine there might be a way of adaptive opacity based on the density of a cluster, but that better be addressed as a separate issue.

**Comment by etpinard** [comment]
> I imagine there might be a way of adaptive opacity based on the density of a cluster, but that better be addressed as a separate issue.

Ok, but why does a selection fix affect panning?

**Comment by dy** [comment]
Slow panning is the direct result of increased number of points caused by #2334 fix. Although it should not be significantly slower than snap-points-2d in 1.35, but possibly can. I did not notice fps drop though. That is the result of z-ordering a point-cluster tree instead of sorting by x, which saves ~100ms of sorting routine on init by displaying extra 10-15% points in runtime.

**Comment by etpinard** [comment]
Thanks for the info @dy !

Can you describe what will happen in turns of performance improvement once (if :wink: ) https://github.com/regl-project/regl/issues/474  is fixed?

**Comment by dy** [comment]
@etpinard 1k 1-point scattergl traces will be rendered ~6 times faster. If number of points per trace more than 1, it is not that noticeable.

**Comment by etpinard** [comment]
Ok. Is there anything we can to improve panning performance at 1e6 pts?

**Comment by etpinard** [comment]
Off this branch, zooming in repetitively

peek 2018-05-01 14-32

leads to:

image

**Comment by etpinard** [comment]
Very strong panning perf off dbd6049:

peek 2018-05-01 17-24

the bug from https://github.com/plotly/plotly.js/pull/2593#issuecomment-385750549 appears to be fixed too :tada:

File: test/jasmine/tests/splom_test.js [hunk 1]
```diff
@@ -549,6 +549,21 @@ describe('@gl Test splom interactions:', function() {
         .catch(failTest)
         .then(done);
     });
+
+    it('should work with typed arrays', function(done) {
```
Comment (etpinard) [inline test/jasmine/tests/splom_test.js]: locking a fix for https://github.com/plotly/plotly.js/issues/2585 from the new `regl-splom` version

**Comment by etpinard** [comment]
All right. Tests are :white_check_mark: Let's merge this thing! :dancer: 

:speaker: Big shoutout to @dy for closing 5 issues in one PR. Huge! :speaker: 

1.37.0 will be out in the next few hours.

**Pull request merged**